src/main/java/com/haxwell/apps/quizki/services/UserServiceImpl.java

```diff
@@ -0,0 +1,125 @@
+package com.haxwell.apps.quizki.services;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import com.haxwell.apps.quizki.dtos.CreatedUserDTO;
+import com.haxwell.apps.quizki.dtos.UserCreateDTO;
+import com.haxwell.apps.quizki.entities.User;
+import com.haxwell.apps.quizki.entities.UserRole;
+import com.haxwell.apps.quizki.exceptions.UserFieldsNotUniqueException;
+import com.haxwell.apps.quizki.exceptions.UserRoleNotInDatabaseException;
+import com.haxwell.apps.quizki.exceptions.ValidationErrorData;
+import com.haxwell.apps.quizki.repositories.UserRepository;
+import com.haxwell.apps.quizki.repositories.UserRoleRepository;
+
+@Service
+public class UserServiceImpl implements UserService {
+	
+	@Autowired
+	private UserRepository uRepo;
+	@Autowired
+	private UserRoleRepository uroleRepo;
+	
+	
+	public UserServiceImpl(UserRepository uRepo, UserRoleRepository uroleRepo) {
+
+		this.uRepo = uRepo;
+		this.uroleRepo = uroleRepo;
+	}
+	
+	
+	public CreatedUserDTO  createNewUser(UserCreateDTO ucdto) {
+		
+		long nameCount = uRepo.countByName(ucdto.getName());
+		long emailCount = uRepo.countByEmail(ucdto.getEmail());
+		
+		if (emailCount != 0 || nameCount != 0) {
+			ValidationErrorData data = new ValidationErrorData();
+			if(emailCount != 0)
+				data.addFieldError("email", "email.not.unique");
+			if(nameCount != 0)
+				data.addFieldError("name", "name.not.unique");
+			
+			
+			throw new UserFieldsNotUniqueException(data);
+		}
+		
+		long role_id = ucdto.getRoleId();
+		Optional<UserRole> urole = uroleRepo.findById(role_id);
+		
+		if(!urole.isPresent()) {
+			ValidationErrorData data = new ValidationErrorData();
+			data.addFieldError("role", "role.not.in.database");
+			throw new UserRoleNotInDatabaseException(data);
+		}
+		
+		//Create a new User instance and map it from the ucdto DTO here
+		User user = new User();
+		
+		user.setName(ucdto.getName());
+		user.setEmail(ucdto.getEmail());
+		user.setRole(urole.get());
+		user.setDemographic("default");				//TODO remove
+		user.setEnabled(1);							//TODO remove
+		user.setFullname(ucdto.getFullname());
+		user.setPassword(ucdto.getPassword());
+		user.setId(null);
+		
+		User savedusr = uRepo.save(user);
+		
+		//then create an instance of CreatedUserDTO and map it from the savedusr
+		//and return the CreatedUserDTO
+		
+		CreatedUserDTO cudto = new CreatedUserDTO();
+		
+		cudto.setEmail(savedusr.getEmail());
```

Comment:
Here too... (call set()'s in the constructor)

Reply:
Obviously the same comments regarding the service layer apply here too. 

As for returning a `User` I didn't want to return the `Demographic`, `Enabled`, `User_Role` and `Password` in the returned object. The first three because the frontend doesn't need to know about them and the last for security.  A DTO allows the response to be uncoupled from the model like this and for the controller to handle the response with the `@Response` annotation. Also I wanted there to be a single place in the code where the front-end developers would be able to explicitly see the objects passed in both directions for reasons we discussed at the meeting on Monday.

I think there could be changes in the frontend-backend interaction again and this architecture gives us good flexibility to handle that. It actually came about because the original way of doing things wasn't flexible enough! For example we'll need to handle changing or resetting passwords and I think that could require keeping track of when a reset password was sent for automatic expiration. The DTO&Service layer would allow us to do that without passing the created date for example.